src/test/coins_tests.cpp

```diff
@@ -585,314 +624,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
     }
 }
 
-static size_t InsertCoinsMapEntry(CCoinsMap& map, CoinsCachePair& sentinel, CAmount value, char flags)
+static size_t InsertCoinsMapEntry(CCoinsMap& map, CoinsCachePair& sentinel, const CoinEntry& cache_coin)
 {
-    if (value == ABSENT) {
-        assert(flags == NO_ENTRY);
-        return 0;
-    }
-    assert(flags != NO_ENTRY);
     CCoinsCacheEntry entry;
-    SetCoinsValue(value, entry.coin);
-    auto inserted = map.emplace(OUTPOINT, std::move(entry));
-    assert(inserted.second);
-    inserted.first->second.AddFlags(flags, *inserted.first, sentinel);
-    return inserted.first->second.coin.DynamicMemoryUsage();
+    SetCoinsValue(cache_coin.value, entry.coin);
+    auto [iter, inserted] = map.emplace(OUTPOINT, std::move(entry));
+    assert(inserted);
+    if (cache_coin.IsDirty()) CCoinsCacheEntry::SetDirty(*iter, sentinel);
+    if (cache_coin.IsFresh()) CCoinsCacheEntry::SetFresh(*iter, sentinel);
+    return iter->second.coin.DynamicMemoryUsage();
 }
 
-void GetCoinsMapEntry(const CCoinsMap& map, CAmount& value, char& flags, const COutPoint& outp = OUTPOINT)
+static MaybeCoin GetCoinsMapEntry(const CCoinsMap& map, const COutPoint& outp = OUTPOINT)
 {
-    auto it = map.find(outp);
-    if (it == map.end()) {
-        value = ABSENT;
-        flags = NO_ENTRY;
-    } else {
-        if (it->second.coin.IsSpent()) {
-            value = SPENT;
-        } else {
-            value = it->second.coin.out.nValue;
-        }
-        flags = it->second.GetFlags();
-        assert(flags != NO_ENTRY);
+    if (auto it{map.find(outp)}; it != map.end()) {
+        return CoinEntry{
+            it->second.coin.IsSpent() ? SPENT : it->second.coin.out.nValue,
+            CoinEntry::ToState(it->second.IsDirty(), it->second.IsFresh())};
     }
+    return MISSING;
 }
 
-void WriteCoinsViewEntry(CCoinsView& view, CAmount value, char flags)
+static void WriteCoinsViewEntry(CCoinsView& view, const MaybeCoin& cache_coin)
 {
     CoinsCachePair sentinel{};
     sentinel.second.SelfRef(sentinel);
     CCoinsMapMemoryResource resource;
     CCoinsMap map{0, CCoinsMap::hasher{}, CCoinsMap::key_equal{}, &resource};
-    auto usage{InsertCoinsMapEntry(map, sentinel, value, flags)};
+    auto usage{cache_coin ? InsertCoinsMapEntry(map, sentinel, *cache_coin) : 0};
     auto cursor{CoinsViewCacheCursor(usage, sentinel, map, /*will_erase=*/true)};
     BOOST_CHECK(view.BatchWrite(cursor, {}));
 }
 
 class SingleEntryCacheTest
 {
 public:
-    SingleEntryCacheTest(CAmount base_value, CAmount cache_value, char cache_flags)
+    SingleEntryCacheTest(CAmount base_value, const MaybeCoin& cache_coin)
     {
-        WriteCoinsViewEntry(base, base_value, base_value == ABSENT ? NO_ENTRY : DIRTY);
-        cache.usage() += InsertCoinsMapEntry(cache.map(), cache.sentinel(), cache_value, cache_flags);
+        auto base_cache_coin{base_value == ABSENT ? MISSING : CoinEntry{base_value, CoinEntry::State::DIRTY}};
```

Comment:
nit: Going from `DIRTY` to `CoinEntry::State::DIRTY` is a bit noisy. I'd prefer making `CoinEntry::State` a non-`class` `enum`.
```suggestion
        auto base_cache_coin{base_value == ABSENT ? MISSING : CoinEntry{base_value, CoinEntry::DIRTY}};
```

Reply:
enum class is safer than enum https://agrawalsuneet.github.io/blogs/enum-vs-enum-class-in-c++ and I didn't like the result of `using enum State`